### PR TITLE
Remove usage of method placeholder functions

### PR DESCRIPTION
**Pull request by devennavani**
This PR introduces a new minor client version `0.67.x` that does not create method placeholder functions for modal classes.

I've tested all forward lookup scenarios (`.lookup()` of a `0.67` class from a pre `0.67` client) as well as backwards lookup scenarios (`.lookup()` of a pre `0.67` class from a `0.67` client) and all work, except for a `0.62` client looking up a `0.67` class, which will be explicitly not supported (this maintains our current restriction of not being able to lookup a `0.63+` class from a `0.62` client).

## Changelog

New minor client version `0.67.x` comes with an internal data model change for how Modal creates functions for Modal classes. There are no breaking or backwards-incompatible changes with this new minor version. All forward lookup scenarios (`.lookup()` of a `0.67` class from a pre `0.67` client) as well as backwards lookup scenarios (`.lookup()` of a pre `0.67` class from a `0.67` client) work, except for a `0.62` client looking up a `0.67` class (this maintains our current restriction of not being able to lookup a `0.63+` class from a `0.62` client).

### CODE REVIEW

File: modal/functions.py [hunk 1]
```diff
@@ -364,94 +363,6 @@ def _bind_method(
         fun._is_method = True
         return fun
 
-    def _bind_method_old(
```
Comment (freider) [inline modal/functions.py:367]: 😍

**Review by freider — COMMENTED** [review]
lovely! did we decide to not backfill method metadata on existing classes in order to remove existing placeholders? (completely understandable, seems messy).

Another related question:
Will "forward" lookups work after this? I.e. can a 0.63 client `.lookup()` a 0.67 class? The way I understand the ClassGet backend implementation the answer is yes, since it "mocks out" placeholders with the function id set to the class service function?

**Comment by devennavani** [comment]
> lovely! did we decide to not backfill method metadata on existing classes in order to remove existing placeholders? (completely understandable, seems messy).
> 
> Another related question: Will "forward" lookups work after this? I.e. can a 0.63 client `.lookup()` a 0.67 class? The way I understand the ClassGet backend implementation the answer is yes, since it "mocks out" placeholders with the function id set to the class service function?

@freider  yes, forward lookups will work, excluding lookups from a `0.62` client, which is currently not supported either.

re: your first question, yea it feels a bit messy. i'm guessing i would have to have the server modify the class service function definitions coming from older clients to include the method definitions mapping and that will likely introduce more confusion

**Comment by devennavani** [comment]
@prbot approve

File: modal/cls.py [hunk 2]
```diff
@@ -244,7 +244,7 @@ class _Cls(_Object, type_prefix="cs"):
     _class_service_function: Optional[
         _Function
     ]  # The _Function serving *all* methods of the class, used for version >=v0.63
-    _method_functions: Dict[str, _Function]  # Placeholder _Functions for each method
+    _method_functions: Dict[str, _Function] = None  # Placeholder _Functions for each method
```
Comment (mwaskom) [inline modal/cls.py:247]: A little ugly to be setting one default value in the middle of a bunch of type annotations

File: modal/cli/import_refs.py [hunk 3]
```diff
@@ -110,7 +110,7 @@ def get_by_object_path(obj: Any, obj_path: str) -> Optional[Any]:
 def _infer_function_or_help(
     app: App, module, accept_local_entrypoint: bool, accept_webhook: bool
 ) -> Union[Function, LocalEntrypoint]:
-    function_choices = set(tag for tag, func in app.registered_functions.items() if not func.info.is_service_class())
+    function_choices = set(tag for tag in app.registered_functions.keys())
```
Comment (mwaskom) [inline modal/cli/import_refs.py:113]: Nit (not worth fixing just FYI) can do this as `set(app.registered_functions)` or just do set operations on the `.keys()` directly.

File: modal/_runtime/container_io_manager.py [hunk 4]
```diff
@@ -452,7 +452,7 @@ async def _dynamic_concurrency_loop(self):
             await asyncio.sleep(DYNAMIC_CONCURRENCY_INTERVAL_SECS)
 
     async def get_app_objects(self) -> RunningApp:
-        req = api_pb2.AppGetObjectsRequest(app_id=self.app_id, include_unindexed=True)
+        req = api_pb2.AppGetObjectsRequest(app_id=self.app_id, include_unindexed=True, only_class_function=True)
```
Comment (mwaskom) [inline modal/_runtime/container_io_manager.py:455]: Do we have a comment explaining the purpose / limited lifetime of `only_class_function` somewhere? I didn't see it in api.proto. (In general we're not very good about documenting the protos, but should try to get better).

File: modal/cli/run.py [hunk 5]
```diff
@@ -136,20 +136,33 @@ def _get_clean_app_description(func_ref: str) -> str:
 
 
 def _get_click_command_for_function(app: App, function_tag):
-    function = app.registered_functions[function_tag]
+    function = app.registered_functions.get(function_tag, None)
```
Comment (mwaskom) [inline modal/cli/run.py:139]: Super nit (not worth fixing) the `None` is superfluous here, that's the default behavior for `.get()`

File: modal/cls.py [hunk 6]
```diff
@@ -273,27 +272,49 @@ def _get_partial_functions(self) -> Dict[str, _PartialFunction]:
...
-                # since each function will already be a loaded dependency _Function
-                self._method_functions[method.function_name]._hydrate(
-                    method.function_id, self._client, method.function_handle_metadata
-                )
+        if (
+            self._class_service_function
+            and self._class_service_function._method_handle_metadata
+            and len(self._class_service_function._method_handle_metadata)
+        ):
+            # The class only has a class service service function and no method placeholders.
```
Comment (mwaskom) [inline modal/cls.py:280]: Is this all classes v0.67+?

File: modal/cls.py [hunk 7]
```diff
@@ -273,27 +272,49 @@ def _get_partial_functions(self) -> Dict[str, _PartialFunction]:
...
+                )
+        else:
+            # pre 0.63 class that does not have a class service function and only method functions
+            self._method_functions = {}
+            for method in metadata.methods:
                 self._method_functions[method.function_name] = _Function._new_hydrated(
                     method.function_id, self._client, method.function_handle_metadata
                 )
 
     def _get_metadata(self) -> api_pb2.ClassHandleMetadata:
```
Comment (mwaskom) [inline modal/cls.py:316]: I'm missing why we have this method that returns an empty proto message now.

File: modal/cls.py [hunk 8]
```diff
@@ -327,56 +348,43 @@ def from_local(user_cls, app: "modal.app._App", class_service_function: _Functio
...
+                app_id=resolver.app_id, existing_class_id=existing_object_id, only_class_function=True
+            )
             resp = await resolver.client.stub.ClassCreate(req)
-            # Even though we already have the function_handle_metadata for this method locally,
-            # The RPC is going to replace it with function_handle_metadata derived from the server.
-            # We need to overwrite the definition_id sent back from the server here with the definition_id
-            # previously stored in function metadata, which may have been sent back from FunctionCreate.
-            # The problem is that this metadata propagates back and overwrites the metadata on the Function
-            # object itself. This is really messy. Maybe better to exclusively populate the method metadata
-            # from the function metadata we already have locally? Really a lot to clean up here...
```
Comment (mwaskom) [inline modal/cls.py:368]: Thanks for cleaning this up 😂

**Review by mwaskom — APPROVED** [review]
I know this mostly involves internal changes but is there anything we should put in a public changelog here? Users will see the version bump and wonder what's non-compatible.

I left a few other comments but mostly Python trivia that you don't need to address.

This change is touching some very complicated stuff and I didn't fully pore through all the details, as I trust you've exercised it thoroughly in your testing. LMK if you'd feel more comfortable with me doing that before you merge.

File: modal/cls.py (hunk 6 above)
Comment (devennavani) [reply modal/cls.py:280]: yes, updated comment

File: modal/cls.py (hunk 7 above)
Comment (devennavani) [reply modal/cls.py:316]: removed

**Pull request merged**